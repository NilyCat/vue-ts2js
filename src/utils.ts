import * as mkdirp from 'mkdirp'

export { readFileSync, writeFileSync, existsSync } from 'fs'
export { resolve, dirname } from 'path'
export const mkdirpSync = mkdirp.sync
export const cwd = () => process.cwd()
