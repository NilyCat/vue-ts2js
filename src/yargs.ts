import * as yargs from 'yargs'

const args = yargs
  // @ts-ignore
  .help(false)
  // @ts-ignore
  .version(false)

export function getYArgs() {
  return args
}

export function baseOptions(yargs: yargs.Argv): yargs.Argv {
  return yargs
    .option('directory', {
      demand: true,
      alias: 'd',
      describe: 'Vue files directory',
      type: 'string'
    })
    .option('output', {
      demand: true,
      alias: 'o',
      describe: 'Output directory',
      type: 'string'
    })
    .option('project', {
      demand: true,
      alias: 'p',
      describe: "TypeScript configuration file, eg:'tsconfig.json'.",
      type: 'string'
    })
}
