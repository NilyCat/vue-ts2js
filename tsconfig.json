{
  "compilerOptions": {
    "experimentalDecorators": true,
    "outDir": "./lib",
    "target": "es2017",
    "module": "commonjs",
    "forceConsistentCasingInFileNames": true,
    "esModuleInterop": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "strict": true,
    "removeComments": true,
    "allowSyntheticDefaultImports": true,
    "listEmittedFiles": true,
    "listFiles": true,
    "allowJs": false,
    "declaration": true,
    "declarationDir": "lib/types",
    "sourceMap": true,
    "importHelpers": true,
    "resolveJsonModule": false,
    "types": [
      "jest",
      "@types/node"
    ]
  },
  "exclude": [
    "coverage",
    "lib",
    "node_modules"
  ]
}
