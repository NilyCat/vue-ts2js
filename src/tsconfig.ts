import { isValid } from '@nily/utils'
import { CompilerOptions } from 'typescript'
import { dirname, resolve } from './utils'

export function compilerOptions(path: string): CompilerOptions {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const content: any = require(path)

  if (isValid(content.extends)) {
    let p = dirname(path)
    p = resolve(p, content.extends)

    return {
      ...compilerOptions(p),
      ...content.compilerOptions,
      sourceMap: false
    }
  } else {
    return {
      ...content.compilerOptions,
      sourceMap: false
    }
  }
}
