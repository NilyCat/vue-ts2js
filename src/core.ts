import { readFileSync } from 'fs'
import globby from 'globby'
import { CompilerOptions, transpile } from 'typescript'
// @ts-ignore
import vueSfcToString from 'vue-sfc-descriptor-to-string'
import { parseComponent } from 'vue-template-compiler'
import { compilerOptions } from './tsconfig'
import { cwd, dirname, mkdirpSync, resolve, writeFileSync } from './utils'

export interface Configs {
  directory: string
  project: string
  output: string
}

export async function process(configs: Configs): Promise<void> {
  const dir = resolve(cwd(), configs.directory)
  const output = resolve(cwd(), configs.output)

  const paths = await globby(dir, {
    onlyFiles: true,
    expandDirectories: {
      extensions: ['vue']
    }
  })

  const options = compilerOptions(resolve(cwd(), configs.project))

  paths.forEach(path => {
    const filePath = path.replace(dir, output)
    const fileDir = dirname(filePath)
    const content = compile(path, options)

    // Create directory
    mkdirpSync(fileDir)

    writeFileSync(filePath, content)
  })
}

export function compile(path: string, options: CompilerOptions): string {
  const content = readFileSync(path).toString('utf-8')
  const sfc = parseComponent(content)

  if (sfc.script && sfc.script.lang === 'ts') {
    sfc.script.lang = 'js'
    sfc.script.attrs.lang = 'js'
    sfc.script.content = ['\n', transpile(sfc.script.content, options)].join('')
  }

  return vueSfcToString(sfc)
}
