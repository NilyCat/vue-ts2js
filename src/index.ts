#!/usr/bin/env node

import { process } from './core'
import * as yargs from 'yargs'

const configs: any = yargs
  .help(false)
  .version(false)
  .usage('Usage: $0 -d [dir] -p [tsconfig.json]')
  .option('directory', {
    demand: true,
    alias: 'd',
    describe: 'Vue files directory',
    type: 'string'
  })
  .option('output', {
    demand: true,
    alias: 'o',
    describe: 'Output directory',
    type: 'string'
  })
  .option('project', {
    demand: true,
    alias: 'p',
    describe: "TypeScript configuration file, eg:'tsconfig.json'.",
    type: 'string'
  })
  .wrap(yargs.terminalWidth())
  .recommendCommands().argv

process(configs)
