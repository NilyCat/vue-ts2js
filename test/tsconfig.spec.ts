import { cwd, resolve } from '../src/utils'
import { compilerOptions } from '../src/tsconfig'

describe('tsconfig', () => {
  test('compilerOptions', () => {
    expect(compilerOptions(resolve(cwd(), 'tsconfig.json'))).toMatchSnapshot()
    expect(compilerOptions(resolve(cwd(), 'tsconfig-build.json'))).toMatchSnapshot()
  })
})
