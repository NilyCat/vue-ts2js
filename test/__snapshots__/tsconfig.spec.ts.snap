// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tsconfig compilerOptions 1`] = `
Object {
  "allowJs": false,
  "allowSyntheticDefaultImports": true,
  "declaration": true,
  "declarationDir": "lib/types",
  "esModuleInterop": true,
  "experimentalDecorators": true,
  "forceConsistentCasingInFileNames": true,
  "importHelpers": true,
  "listEmittedFiles": true,
  "listFiles": true,
  "module": "commonjs",
  "noUnusedLocals": true,
  "noUnusedParameters": true,
  "outDir": "./lib",
  "removeComments": true,
  "resolveJsonModule": false,
  "sourceMap": false,
  "strict": true,
  "target": "es2017",
  "types": Array [
    "jest",
    "@types/node",
  ],
}
`;

exports[`tsconfig compilerOptions 2`] = `
Object {
  "allowJs": false,
  "allowSyntheticDefaultImports": true,
  "declaration": true,
  "declarationDir": "lib/types",
  "esModuleInterop": true,
  "experimentalDecorators": true,
  "forceConsistentCasingInFileNames": true,
  "importHelpers": true,
  "listEmittedFiles": true,
  "listFiles": true,
  "module": "commonjs",
  "moduleResolution": "node",
  "noUnusedLocals": true,
  "noUnusedParameters": true,
  "outDir": "./lib",
  "removeComments": true,
  "resolveJsonModule": false,
  "skipLibCheck": true,
  "sourceMap": false,
  "strict": true,
  "target": "es2015",
  "types": Array [
    "jest",
    "@types/node",
  ],
}
`;
